codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -508,51 +392,49 @@ private void insertHeader(Http2Headers headers, CharSequence name, CharSequence
             default:
                 throw new IllegalStateException("should not reach here");
         }
-    }
 
-    private void addHeader(Http2Headers headers, CharSequence name, CharSequence value) {
-        long newSize = headerSize + name.length() + value.length();
-        if (newSize <= maxHeaderSize) {
-            headers.add(name, value);
-            headerSize = (int) newSize;
-        } else {
-            // truncation will be reported during endHeaderBlock
-            headerSize = maxHeaderSize + 1;
-        }
+        return headerSize;
     }
 
-    private boolean exceedsMaxHeaderSize(long size) {
-        // Check new header size against max header size
-        if (size + headerSize <= maxHeaderSize) {
-            return false;
+    private int addHeader(Http2Headers headers, CharSequence name, CharSequence value, int headersLength)
+            throws Http2Exception {
+        long newHeadersLength = headersLength + name.length() + value.length();
+        if (newHeadersLength > maxHeadersLength) {
+            maxHeaderSizeExceeded();
         }
+        headers.add(name, value);
+        return (int) newHeadersLength;
```

Comment:
I don't think so because we do bounds checking a few lines above `if (newHeadersLength > maxHeadersLength) {` (`maxHeadersLength` is a `int` and we are using `long` arithmetic for the addition of the 3 ints which we compare it against).